airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py

```diff
@@ -0,0 +1,115 @@
+
+    def _parse_options(self) -> Mapping[str, str]:
+        """
+        https://arrow.apache.org/docs/python/generated/pyarrow.json.ParseOptions.html
+         build ParseOptions object like: pa.json.ParseOptions(**self._parse_options())
+        """
+
+        return {"newlines_in_values": self.format.newlines_in_values, "unexpected_field_behavior": self.format.unexpected_field_behavior}
+
+    def get_inferred_schema(self, file: Union[TextIO, BinaryIO]) -> Mapping[str, Any]:
+        """
+        https://arrow.apache.org/docs/python/generated/pyarrow.json.read_json.html
+        This now uses multiprocessing in order to timeout the schema inference as it can hang.
+        Since the hanging code is resistant to signal interrupts, threading/futures doesn't help so needed to multiprocess.
+        https://issues.apache.org/jira/browse/ARROW-11853?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel
+        """
+
+        def infer_schema_process(file_sample: str, read_opts: dict, parse_opts: dict) -> Tuple[dict, Optional[Exception]]:
+            """
+            we need to reimport here to be functional on Windows systems since it doesn't have fork()
+            https://docs.python.org/3.7/library/multiprocessing.html#contexts-and-start-methods
+            This returns a tuple of (schema_dict, None OR Exception).
+            If return[1] is not None and holds an exception we then raise this in the main process.
+            This lets us propagate up any errors (that aren't timeouts) and raise correctly.
+            """
+            try:
+                import tempfile
+
+                import pyarrow as pa
+
+                # writing our file_sample to a temporary file to then read in and schema infer as before
+                with tempfile.TemporaryFile() as fp:
+                    fp.write(file_sample)  # type: ignore[arg-type]
+                    fp.seek(0)
+                    streaming_reader = pa.json.read_json(fp, pa.json.ReadOptions(**read_opts), pa.json.ParseOptions(**parse_opts))
+                    schema_dict = {field.name: field.type for field in streaming_reader.schema}
+
+            except Exception as e:
+                # we pass the traceback up otherwise the main process won't know the exact method+line of error
+                return (None, e)
+            else:
+                return (schema_dict, None)
+
+        # boto3 objects can't be pickled (https://github.com/boto/boto3/issues/678)
+        # and so we can't multiprocess with the actual fileobject on Windows systems
+        # we're reading block_size*2 bytes here, which we can then pass in and infer schema from block_size bytes
+        # the *2 is to give us a buffer as pyarrow figures out where lines actually end so it gets schema correct
+        schema_dict = self._get_schema_dict(file, infer_schema_process)
+        return self.json_schema_to_pyarrow_schema(schema_dict, reverse=True)  # type: ignore[no-any-return]
```

Comment:
I think the complexity of this function comes from a bug related to CSV parsing in PyArrow, I think we can make this function lighter for JSON.

Reply:
I have removed external process as json reader internally read in threads